### ISSUE TITLE
Doubts in coveralls setup for private repo

### ISSUE THREAD

**Issue by indiVar0508** [lemurheavy/coveralls-public#1730]
Hi,

I recently migrated from codecov to coveralls for my orgs private repo, i had some minor doubts which i am not clear yet on

- Sometimes by the time my CI runs, some other code gets merged to main => coverall report is not generated in this case asking me to rebase. This happened multiple times for my team. is there a way to handle this?
           `Target branch is ahead of PR base. Rebase the PR to fix.`

- How does changed tab work for a build, say i have added & modified some filed in PR sometimes the changed tab just shows 0 changed, but in main branch it shows changed files( i assume it from last build's change ) 
- Patch job coverage says 100% coverage is there a way to see just patch changes in build summary in coveralls so i just see changes in build for only changes that i have added

Thanks

**Comment by indiVar0508** [comment]
@afinetooth any suggestion for this one?
`Target branch is ahead of PR base. Rebase the PR to fix.`

**Comment by afinetooth** [comment]
Hi, @indiVar0508. Welcome.

Yes, this is a message we've started posting at the top of `pull_request` build pages to address the fact that, if you're only building `pull_request` commits on your PR branches, we may not be able to deliver a 100% accurate coverage report every time as you'd expect.

This is because GitHub `pull_request` commits use a [two dot diff comparison](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests), which changes the diff when the base branch is updated, vs. a [three dot diff comparison](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests), which is what most users want to see when they're considering the coverage report for a PR. Specifically, the [three dot diff comparison](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#about-three-dot-comparison-on-github) shows _just the changes introduced by the pull request_, and, just the _coverage_ associated with those changes.

When you're only building `pull_request` commits on your PR branches in CI. Per:

```
on:
  push:
    branches: [main]
  pull_request:
    branches: [main]
```

You're only sending Coveralls coverage reports for `pull_request` builds (on your PR branches), and, therefore, it can't make a _guaranteed_ [three dot diff comparison](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#about-three-dot-comparison-on-github). It can only do that if there have been no further commits on your base branch. (Or if you rebase your PR, as the in-page message suggests.)

Once your PR base branch (aka. target branch) is updated, the `pull_request` commit, in providing a [two dot diff comparison](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#two-dot-git-diff-comparison), includes the new changes made to the target branch; so, at that point, the only way Coveralls could provide the expected ([three dot](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#about-three-dot-comparison-on-github)) diff is if it had the two relevant `push` commits from _both_ your PR branch _and_ your target branch.

This is why we've started recommending that you build `push` commits on your PR branches as well, so you can guarantee that Coveralls can make a [three dot comparison](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#about-three-dot-comparison-on-github), which always provides an accurate coverage report for your PRs.

```
on:
  push
```

This configuration guarantees that Coveralls can always find the two commits that represent "[just what the PR introduces](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#about-three-dot-comparison-on-github:~:text=what%20a%20pull%20request%20introduces)."

**What about PR Builds, though?**:
In addition to delivering the above warning message, Coveralls now renders any `push` build that _belongs_ to a `pull_request` _as_ a **Pull Request Build**. This means that the build will be represented as a PR in the Coveralls UI, with the correct PR #, a link to the PR from the build, and a link back to PR's base build. Basically, the same PR Builds you could previously expect when sending us coverage reports from your `pull_request` builds; but, because we have the `push` builds, we can always guarantee the right diff and the right _coverage_ diff for your PRs.

Hope that all makes sense. Please let me know if not. We are still dialing in some details of the user experience including the messaging we deliver with this new approach. We'll soon start updating documentation with our associated recommendations ("Build `push` commits, or rebase your PRs."). We are always open to your feedback because it helps us get these things right.

---

P.S. As for the other two concerns you raised above, they are related, but I will address them in a follow up comment. The above is the foundation for understanding those remaining items.

**Comment by afinetooth** [comment]
@indiVar0508 (Cc: @mrexox) I just noticed that you said this about the cases where you're seeing `Target branch is ahead of PR base. Rebase the PR to fix.`:

> [...] coverall report is not generated in this case

Can you please share an example build where this happened?

I ask because that is not supposed to be the behavior. Instead, Coveralls should generate a build, but deliver the above message as an **in-app notification** at the top of the PR Build Page, delivering that warning and providing options for ensuring an accurate coverage report.

As you have reported it, it sounds like you might be receiving the warning message as an API response from the Coveralls API, is that right? If so, you can please find the Coveralls API response from your CI build log for the relevant build. If you could share that with us, I'd appreciate it so we can find out why that happened. 🙏

**Comment by afinetooth** [comment]
Hi again, @indiVar0508.

Let me address the remaining two concerns you mentioned above:

> - How does changed tab work for a build, say i have added & modified some filed in PR sometimes the changed tab just shows 0 changed, but in main branch it shows changed files( i assume it from last build's change )

(Adding @nebolsin)

I believe we are experiencing some issues in the CHANGED tab right now due to a regression caused by other recent changes. We're working on a fix, but let me at least share the way it has always worked, and _may_ continue working going forward. (The other option on the table is that we may remove this tab, since the constituent elements can be found elsewhere on the build page.):

Up until recently, there were three related tabs in the [SOURCE FILES table](https://docs.coveralls.io/app-ui#build-page:~:text=Source%20Files%20%E2%80%94%20In%20build%3B%20Name%2C%20Coverage%20%25%20%26%20details%3B%20link%20to%20Source%20File%20Pages), which appears at the bottom of each [Coveralls Build Page](https://docs.coveralls.io/app-ui#build-page):

- **CHANGED**
- **SOURCE CHANGED**
- **COVERAGE CHANGED**

I'll explain them out-of-order, since that will make the most sense:

- Coveralls tracks the changes in the coverage reports for each of your source files between any two builds.
- That is to say that, from the difference in the coverage stats we receive for each of your source files (we do not store source code, ever), we are able to tell if any lines of the source file changed, and what the coverage impact was.
- Therefore, we are able to deliver a list of source files with source code changes in the **SOURCE CHANGED** tab.
- And we are able to deliver of list of files where _coverage_ has changed under the **COVERAGE CHANGED** tab.
- Finally, we deliver the union of files in the **SOURCE CHANGED** and **COVERAGE CHANGED** tabs in the more generally named **CHANGED** tab.

> - Patch job coverage says 100% coverage is there a way to see just patch changes in build summary in coveralls so i just see changes in build for only changes that i have added

Since this sounds like a potential feature request to me, please see my **questions below** about that, if they're still relevant after this explanation:

If you refer to the above explanation about how Coveralls aims to provide a [three dot diff comparison](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#about-three-dot-comparison-on-github) for each of your pull request builds, you'll understand what I mean when I say that your "patch" _is_ that [three dot diff](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#about-three-dot-comparison-on-github). And the **patch status update** you receive from your **Coveralls PR Build** is a coverage report for just that diff (just the _changes_ made in your PR).

We highlight _aspects_ of those changes in several ways on our **PR Build Pages**, which include: 

- The lists of files in the SOURCE CHANGED and COVERAGE CHANGED tabs, which represent the total files altered by the PR in any way.
- PR DIFFS, which are missed coverage opportunities, of which we show two types: (1) NEW UNCOVERED LINES: and (2) UNCOVERED EXISTING LINES. Those lines are annotated in the margins of each source file view with the "UNCOV" label. 

But we don't specifically isolate all the changes in the PR in a single view, since we figure you can get that elsewhere (GitHub), and we decided the most helpful thing to show would be all actionable "coverage opportunities" represented by those "PR DIFFS."

If you want to see the "diff" for a given file, the best thing to do is compare the same file, across builds, side-by-side. Here are the steps:

1. In your **PR Build**, where you're viewing a source file, click on the **PR Base Build** link in the inner left nav.
2. Open the **Base Build** in a new browser tab, and drag that tab out and to the **left** of your screen, so it sits along side your **PR Build** source file view on the **right**.
3. Find the same **source file** you're viewing on the **right**, in the SOURCE FILES table of your **Base Build** on the **left** (The easiest way is usually to search in the LIST view of the SOURCE FILES table).
4. Compare the two versions of the file (Base Build-to-PR Build) side-by-side to see both source and coverage changes.

Here's an example from our (public) Coverage Reporter repo:
https://coveralls.io/github/coverallsapp/coverage-reporter

Recent PR Build (RIGHT):
https://coveralls.io/builds/62381149 ([Source file view for /src/coverage_reporter/parser.cr](https://coveralls.io/builds/62381149/source?filename=src%2Fcoverage_reporter%2Fparser.cr))

That PR's Base Build (LEFT):
https://coveralls.io/builds/62349781 ([Source file view for /src/coverage_reporter/parser.cr](https://coveralls.io/builds/62349781/source?filename=src%2Fcoverage_reporter%2Fparser.cr))

<img width="2558" alt="Screenshot 2023-09-06 at 4 00 39 PM" src="https://github.com/lemurheavy/coveralls-public/assets/918667/85a55100-cf92-478e-ba5f-6f77a73d7c58">

---

**Questions re: potential feature request**:
I sense a feature request, maybe something you used at Codecov?:
1. Would you find value in a GitHub PR-style diff view in the Coveralls PR Build Page UI, showing a summary of source and coverage changes; and
2. How would that ideally look to you? Like a GitHub PR Diff View but with coverage overlaid? Or something else? (Feel free to provide a screenshot or link, or just a description.) 🙏

**Comment by indiVar0508** [comment]
Hi @afinetooth, 

Thanks for the patient & comprehensive explanations. it was really helpful and insightful!

I will start with this one first.

`Target branch is ahead of PR base. Rebase the PR to fix. `

Yes, unfortunately we don't have push event build for now, as our CI tests takes time (almost 2 hours for as of now), so rather we have daily tests on our default branch.

However, the reports are not being generated i have reproduced same in a dummy project i have: https://github.com/indiVar0508/Dummy/pull/20
In the Build checks if i open the details tab it is empty

![image](https://github.com/lemurheavy/coveralls-public/assets/23294511/6eeac40b-ad3d-48f0-ab41-7d4a9c0d8f28)

In my Private Repo i also have parallel build that do not generate reports saying it is waiting on coveralls.io
![image](https://github.com/lemurheavy/coveralls-public/assets/23294511/9fc2008e-f3b6-4100-b218-379feb20385e)

Even in this i mainly care about the patch job from coveralls to pass but ig it goes back to three-dot-diff, so it can't have any workaround apart from push event on main branch? like something like on merge event take the coverage data for last build from PR and merge it with base branch coverage data ( i am sorry i don't know much about coverage internals just shooting in the dark ).

`Would you find value in a GitHub PR-style diff view in the Coveralls PR Build Page UI, showing a summary of source and coverage changes` - Yeah sure, that will be helpful!

`How would that ideally look to you? Like a GitHub PR Diff View but with coverage overlaid? Or something else? (Feel free to provide a screenshot or link, or just a description.)` - I personally i haven't worked on CodeCov much but a colleague in my org has worked with it where he mentioned codecov used to give a diff + coverage like mentioned in this one https://app.codecov.io/gh/pandas-dev/pandas/pull/55062 (expand any file to see the diff)

![image](https://github.com/lemurheavy/coveralls-public/assets/23294511/c1aa356f-7936-4570-a934-bae555a7bc25)


Another query that I had was regarding pull request comment setup.

so I noticed that I added coveralls-bot as collaborator for my project (even for the dummy project mentioned above).
Now if I check it is no longer present as collaborator in my project.

1) Is this ok, like once bot accepts initiation it removes itself?
2) for my private repo, i have to pay per user in my org so if I add coveralls-bot with READ access do i have to pay for it separately?.

**Comment by afinetooth** [comment]
Hi @indiVar0508 

> `Target branch is ahead of PR base. Rebase the PR to fix. `
> 
> Yes, unfortunately we don't have push event build for now, as our CI tests takes time (almost 2 hours for as of now), so rather we have daily tests on our default branch.

Hmm. Yes, I understand. Let me think about any recommendations there.
 
> However, the reports are not being generated i have reproduced same in a dummy project i have: [indiVar0508/Dummy#20](https://github.com/indiVar0508/Dummy/pull/20) In the Build checks if i open the details tab it is empty
> 
> ![image](https://user-images.githubusercontent.com/23294511/266637937-6eeac40b-ad3d-48f0-ab41-7d4a9c0d8f28.png)

Yes, I noticed that. Thanks for creating a test. I think we have a (separate) issue there because your coverage report looks fine to me. I have asked an engineer to look at it.

That said, perhaps we can factor this test build out for now because I do see the reason for you not receiving PR Comments or Status Updates there. It is because [that build, associated with PR #20](https://coveralls.io/builds/62527656), cannot find it's base build, whose diff with the current commit is required to derive the contents of both PR Comments and Status Updates. You can tell this is the case when the build reads "FIRST BUILD", but you know it's a build on a PR branch, which definitely has a base commit (the base of the PR on the target branch).

The **base build** in this case should be for commit sha: `c4db4c16925a39b1c16eca34a20f6459a3ef2909`

Per this PR info:

```
{
  "title": "2",
  "head": {
    "sha": "7183f6e28289c05d54589615e80a58a21c6348cb",
    "label": "indiVar0508:poc_2",
    "ref": "poc_2"
  },
  "base": {
    "sha": "c4db4c16925a39b1c16eca34a20f6459a3ef2909",
    "label": "indiVar0508:master",
    "ref": "master"
  }
}

```

But while you have at least one build on `master` for this repo, preceding the PR build, it is apparently not the correct base commit, as its commit sha is `1daf6ee828c0590090608ab2612f864824a64c8b`, and Coveralls cannot find any build with the correct commit sha:

```
repo.builds.where(commit_sha: "c4db4c16925a39b1c16eca34a20f6459a3ef2909").count
=> 0
```

<img width="976" alt="Screenshot 2023-09-12 at 5 40 12 PM" src="https://github.com/lemurheavy/coveralls-public/assets/918667/116ada9d-107e-480d-86de-53e5d86055ff">
 
> In my Private Repo i also have parallel build that do not generate reports saying it is waiting on coveralls.io ![image](https://user-images.githubusercontent.com/23294511/266638532-9fc2008e-f3b6-4100-b218-379feb20385e.png)

Is this the same repo as above? It doesn't look like it because the builds for [indiVar0508/Dummy](https://coveralls.io/github/indiVar0508/Dummy) don't appear to be parallel.

Can you please share the URL for that repo? (I will have a look.)

In the meantime, do you happen to be using the [GitHub Merge Queue](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue) with this repo?
 
> Even in this i mainly care about the patch job from coveralls to pass but ig it goes back to three-dot-diff, so it can't have any workaround apart from push event on main branch? like something like on merge event take the coverage data for last build from PR and merge it with base branch coverage data ( i am sorry i don't know much about coverage internals just shooting in the dark ).

Yes, the workaround is to rebase your PR, meaning do exactly what you said, which is to merge the latest changes on your base branch into your PR branch in a way that replays the PR's changes on top of the base branch changes. Then, for the new PR build, while the diff from GitHub is still a `two-dot-diff`, it is still equivalent to a `three-dot diff` because there are no changes on the base branch that aren't also in the PR.

Here's [how to rebase your PR using GitHub](https://github.blog/changelog/2022-02-03-more-ways-to-keep-your-pull-request-branch-up-to-date/#update-your-pull-request-branch-by-rebasing).

> `Would you find value in a GitHub PR-style diff view in the Coveralls PR Build Page UI, showing a summary of source and coverage changes` - Yeah sure, that will be helpful!

OK! Thank you.

> `How would that ideally look to you? Like a GitHub PR Diff View but with coverage overlaid? Or something else? (Feel free to provide a screenshot or link, or just a description.)` - I personally i haven't worked on CodeCov much but a colleague in my org has worked with it where he mentioned codecov used to give a diff + coverage like mentioned in this one https://app.codecov.io/gh/pandas-dev/pandas/pull/55062 (expand any file to see the diff)
> 
> ![image](https://user-images.githubusercontent.com/23294511/266763338-c1aa356f-7936-4570-a934-bae555a7bc25.png)

OK, thank you. That's very helpful. We have a feature like that planned, but it helps to get an idea of a UX that suits end user needs. We will be doing some testing of the feature later, so I may reach out for input pre-release.

> Another query that I had was regarding pull request comment setup.
> 
> so I noticed that I added coveralls-bot as collaborator for my project (even for the dummy project mentioned above). Now if I check it is no longer present as collaborator in my project.
> 
> 1. Is this ok, like once bot accepts initiation it removes itself?

No, actually. The bot should be a collaborator on your project. I will check to verify it auto-accepted your invitation and get back.

## UPDATE
I checked out bot account and I don't see that it belongs to any org or repos for either `indiVar0508` or `corridor`. Let me know if there's another account I should check.

> 2. for my private repo, i have to pay per user in my org so if I add coveralls-bot with READ access do i have to pay for it separately?.

Yes, I'm afraid this is the case now ay GitHub. I'm not sure if it's always been the case, or if it changed.

**Alternative approach (No bot needed)**:
But in any case, the need for a PR Commenter bot is a requirement of our legacy GitHub integration, which is based on an OAuth App. We will soon replace that with our [New GitHub Integration](coveralls.io/new-github-integration-beta), which is currently in [beta](coveralls.io/new-github-integration-beta). I recommend [joining the beta](coveralls.io/new-github-integration-beta) and migrating to the new integration, which just takes a few seconds to a few minutes and will result in you not having to worry about the bot. PR Comments will just work through the GitHub App.

**Comment by indiVar0508** [comment]
Hi @afinetooth ,

Apologies for getting delayed in response got involved in some fire fighting crisis at work.

Thanks i see some thing has changed it seems since now i don't face this issue in opensource projects that we have and it has helped us a lot.!
`Target branch is ahead of PR base. Rebase the PR to fix.`
Occurrence of this in my private repo has also decreased but it still happens occasionally, one recent example can be seen in this build
`https://coveralls.io/builds/63529189` where coverage was not generated. 

Also just curious, i see sometimes my split coveralls jobs it says it failed but overall pull request job has passed.
eg : https://github.com/indiVar0508/Dummy/pull/27

`I checked out bot account and I don't see that it belongs to any org or repos for either indiVar0508 or corridor`
Ummmm, but we have been getting comments for opensource project that we have https://github.com/corridor/sqlalchemy-history/pull/69, yeah but same doesn't happen for private repos.

**Comment by indiVar0508** [comment]
++ bump

**Comment by indiVar0508** [comment]
Hi @afinetooth , 

Did you get a chance to review this issue?

**Comment by afinetooth** [comment]
Hi @indiVar0508 I wanted to give you an update that we have deployed changes rolling back the the related feature-in-transition to its initial state, which means that we have stopped skipping "out-of-sync" PR builds and are now processing those again, and sending their associates Status Updates and PR Comments. (Here's a [similar notice](https://github.com/lemurheavy/coveralls-public/issues/1733#issuecomment-1792960612) to another customer with a related issue here, for reference.)

That means that you should stop seeing any instances of your builds being paused or skipped, or having any builds for which coverage was not generated.

**Re: your issue**:
Two parts, as far as I can tell:

1. **Re**: `https://coveralls.io/builds/63529189` **where coverage was not generated** - Per my comment above, this should no longer occur.
2. **Re: sometimes my split coveralls jobs it says it failed but overall pull request job has passed** - I'm pretty sure this is because you have coverage threshold settings set for commit statuses in [your repo settings here](https://coveralls.io/github/indiVar0508/Dummy/settings):

<img width="729" alt="Screenshot 2023-11-03 at 12 04 29 PM" src="https://github.com/lemurheavy/coveralls-public/assets/918667/53a5dcc1-9ce9-4df7-8b8a-b3e8446d2910">

Removing those would be the fastest way to ensure that commit statuses for individual jobs don't block your PRs. (I am looking into whether you can separate out those settings by job/subproject.)

**Comment by dgw** [comment]
> I'm pretty sure this is because you have coverage threshold settings set for commit statuses in your repo settings

> Removing those would be the fastest way to ensure that commit statuses for individual jobs don't block your PRs.

Sorry to chime in on someone else's request, but this reminded me of a question my project has had for a while and never got around to asking: Since we want the coverage-change statuses to be informational only (at least for now), we can prevent them from "failing" any PR by just _clearing_ the threshold settings shown above? Whoever set up Coveralls originally (before my time) seems to have just thrown in some guessed values to silence the failures 😅 but just opting out of the unwanted failure thresholds entirely would be even better if it's supported!

**Comment by indiVar0508** [comment]
Hi @afinetooth ,

`Per my comment above, this should no longer occur.`
Thanks!, yeah some of our team mates noticed this couple of days ago that their PR was not getting blocked for being out of sync.
This was really helpful and appreciated 👍.

`I'm pretty sure this is because you have coverage threshold settings set for commit statuses in [your repo settings here](https://coveralls.io/github/indiVar0508/Dummy/settings)`: 
Ahh I see, will check that but just curious these are actions split job that execute subset of tests in each machine. so individually they would not have complete coverage, but its not a big issue for me as our teams focus is mostly on patch coverage.

Thanks again for helping and explanations in this thread it was really helpful!

**Comment by afinetooth** [comment]
@indiVar0508 great, glad to hear that helps. 

I will close this given the main topic, but regarding the threshold settings: Apologies if the those are over-applying to your individual jobs. I have opened an engineering ticket about it and will update you here on status and any changes.

**Comment by indiVar0508** [comment]
Hi @afinetooth ,

The patch job coverage is not getting generated still it seems, 
but now coverage on pull request is generated with apt. error message on being out-of-sync, 
I tried to open the guidelines link but it gives me link not found error
ref job: https://coveralls.io/builds/63772388
![image](https://github.com/lemurheavy/coveralls-public/assets/23294511/8f5fdebd-0fb8-40a6-b7c1-6f00fb3ecb8e)
![image](https://github.com/lemurheavy/coveralls-public/assets/23294511/00237cbc-f236-405a-8bee-0b1eb1ccff61)

**Comment by afinetooth** [comment]
Hi, @indiVar0508.

Yes, apologies. We had an issue getting our docs update out in time. That has been deployed now, though.

One thing to know, however, is that the link in that warning message has changed. It should now link you to here:
https://docs.coveralls.io/build-types#recommended-ci-configurations

I checked it against your build:
https://coveralls.io/builds/63772388

If you refresh that page, it should take you to the right place.

I am looking into your patch status update...

**Update**:
While I can't see it at your pull request at GitHub, it looks like there was a patch status update sent, which would have had the `state: success` and the `description: No relevant coverage`.

I re-processed your build in the hopes that would be re-sent. You'll have to let me know if you see it now. (If not, I'll look into it further.)

**Comment by indiVar0508** [comment]
Hi @afinetooth ,

`You'll have to let me know if you see it now. (If not, I'll look into it further.)`
Yeah it seems it didn't come through for PR after the re-build, i will re-trigger the test from my end without rebasing to see if the issue persists and update here.

**UPDATE 1** 
Re-run the CI still got the empty patch [job](https://coveralls.io/builds/63772388).

![image](https://github.com/lemurheavy/coveralls-public/assets/23294511/76e52b51-77d1-4819-897e-bd01a9739235)


**UPDATE 2**
I rebased and ran the job still the patch job didn't got generated no issue is reported in build job (https://coveralls.io/builds/63804735)
![image](https://github.com/lemurheavy/coveralls-public/assets/23294511/110d5fef-92d1-48cc-9134-61a61ea764c5)



Another issue that we encountered today which seems to be in same line is the status flag behavior based on threshold of patch jobs.
Build ref: https://coveralls.io/builds/63791842

![image](https://github.com/lemurheavy/coveralls-public/assets/23294511/20b436ab-6e7b-494d-87fc-e1cf79b115c2)

![image](https://github.com/lemurheavy/coveralls-public/assets/23294511/515f0c13-89f1-4978-8c16-b448fae1beae)

patch job says the coverage is 98...% but our threshold is 100% so the job status flag should have failed but it came as success.

also if it is fine can we re-open this issue for some days until these issues are resolved or we can open a new issue thread?

**Comment by indiVar0508** [comment]
Hi, 

minor follow up on this, we encountered another issue where a job even though running on latest rebased default branch is not generating patch coverage
https://coveralls.io/builds/63877996
![image](https://github.com/lemurheavy/coveralls-public/assets/23294511/f4a21731-e8cf-4d17-8a3f-d30d8dd3242d)

**Comment by afinetooth** [comment]
Hi @indiVar0508,

For that PR build:
https://coveralls.io/builds/63877996

**Reason for no relevant patch status content**:
Coveralls can't find the PR's base build (the coverage report for the PR's base commit). Without that comparison, Coveralls can't generate the content for a status update about the change in coverage. 

You can tell this is the case when the Build Info table reads "FIRST BUILD", like this:
 
<img width="941" alt="Screenshot 2023-11-13 at 5 02 51 PM" src="https://github.com/lemurheavy/coveralls-public/assets/918667/bd7ff5ec-0491-4763-9073-9481cd2f0a9f">

The base build for a PR build is usually a `push` build on your default branch, but in this case, your build history looks like it has a lot more PR builds than `push` builds, and, for instance, is looking for a `push` build based on a different commit than the one that most recently preceded it:

![screencapture-coveralls-io-repos-411301-builds-2023-11-13-17_07_51](https://github.com/lemurheavy/coveralls-public/assets/918667/c6a7a927-c163-4e0d-9f73-eb54dad2fc00)

Are you aware of any reason you would not have sent a coverage report for the base commit the PR build is looking for? (That commit SHA is: `7b68a9816fe997893375a2ea3ba0864a17d87e41`

**Re: patch status update**:
I actually do see that a patch status update was generated, which reads "No relevant coverage" because Coveralls can't find a coverage report to compare with the HEAD commit of the PR, but I can't confirm whether it was sent. Can you please check that again to see if it ever arrived?

I see that we have a call scheduled for next week to cover this and a few other issues. We can do our best here in the meantime, but we'll try to wrap them all then.
